### ISSUE TITLE
Issue 1138 - Defaults - haxe

### ISSUE THREAD

**Issue by issuesbot** [HaxeFoundation/haxe#1138]
<i>[Google Issue #1138 : https://code.google.com/p/haxe/issues/detail?id=1138]</i>
 <i>by ncanna...@gmail.com, at 09/08/2012, 13:03:21</i>
Have a special file named &quot;defaults.hx&quot; (lowercase to differentiate from normal file ?) which contains :

- type definitions globally accessible
- &quot;import&quot; which are done by default
- &quot;using&quot; which are done by default

This file will apply to the package it is defined into and to all its sub packages. Defaults can shadow each other definitions the normal way.

- a haxelib library can also have its own &quot;defaults.hx&quot; which activates itself when the user link the library with -lib (less priority than project-specific defaults)

- it is possible to specify additional (project-whole) defaults files by commandline (for configuration for instance)

To prevent conflicts :

- defaults of the project does not apply to libraries
- defaults of one library does not apply to other libraries (except if there is a dependency ? - hard to tell/implement)

**Comment by nadako** [comment]
Can we at least get `--macro addGlobalUsing(...)` sometime soon? I feel need for it quite often.

**Comment by ncannasse** [comment]
@nadako no for the same reason : this would affect the scope of your libraries as well. All the global switches that changes the code behavior in uncontrolled ways should be prohibited

**Comment by ncannasse** [comment]
But let's put it on 3.3 milestone for now

**Comment by nadako** [comment]
Just an idea, if we implement typed metadata (see #3959) with metadata macro hooks, we could use that to enable per-package/per-module/per-type `using` of a concrete class, syntatically similar to how `@:allow` is used.

I.e.:
```haxe
@:useIn(myapp) // enable for myapp.*
class MyAppExtensions {
    public static function times(a:Int, b:Int):Int return a * b;
}
```

**Comment by Simn** [comment]
This would introduce order constraints, i.e. MyAppExtensions has to be compiled before any type that might rely on it.

**Comment by nadako** [comment]
I see. Yeah, then I guess `defaults.hx` is the best option. Looking forward for it to be implemented. :-)

**Comment by Simn** [comment]
I'm not sure I like this lowercase file name business. It is already causing problems with our completion. There are issues when you have a `defaults` package next to a `defaults.hx`. But even without that I'm getting defaults reported as available package.

What do we really gain from making this lower-case? It's not like we could have `defaults.hx` next to `Defaults.hx` in a cross-platform way. I think I would prefer if we dealt with #4462 and go for something like `HxDefaults.hx` or whatever. That way it would also be possible to manually import the defaults module like @ousado suggested.

**Comment by Simn** [comment]
Another thing: I have implemented a stupid parser cache so we wouldn't have to parse the file(s) for every module. Reading through #1214 made me realize that this is not safe with regards to conditional compilation. Not sure what to do about that right now, but we definitely need some sort of cache for this.

**Comment by nadako** [comment]
The point about `defaults.hx` vs `Defaults.hx` is very valid on Windows since filesystem is case-insensitive there, however I don't like the idea of having *defaults* importable as a normal module, since it's quite special. How about we name it `_defaults.hx`? We can't have normal module names starting with an underscore, right? One nice bonus is that `_defaults.hx` will show up first in the sorted module list, so one can easily find it. cc @ncannasse

Anyway, this feature is really great for large code bases (with own set of static extensions and special types), so I'm looking forward for it to be merged into development soon!

**Comment by nadako** [comment]
There's also a collective concern about the term "defaults" itself. Wouldn't something like "globals" be more appropriate? However, it's not exactly global as well, as its scope is limited to its directory. Any ideas?

**Comment by back2dos** [comment]
Is this going to be constrained by directories, or by packages? I think the former would help a lot in avoiding any conflicts. You could even have libraries throw this stuff in the root name space and it wouldn't affect anyone.

As for a package-based approach, IIRC I proposed metadata-based imports/usings (that would be scoped to whatever the metadata is applied to) and Simn even had implemented that before it got rejected for somewhat intransparent reasons. If we had it, then `addGlobalMetaData` would give is this feature.

But if we're going with a file based solution, we might as well call it `import.hx` since that's pretty clear. And `import path.pack;` (not the lack of an actual type) could import such a file.

**Comment by nadako** [comment]
It's currently directory-based which is sufficient and less-problematic than package-based IMO.

**Comment by nadako** [comment]
`import.hx` name suggests that it contains only imports, but it actually can contain type definitions (i'm not sure if it's useful though).

**Comment by Simn** [comment]
Let's restrict it to import and using, that takes care of some awkward questions regarding shadowing.

**Comment by nadako** [comment]
Agreed, I'm not sure we want to give users freedom to shoot their own kneecap by allowing type definitions in defaults.hx

**Comment by nadako** [comment]
Then we can indeed name it `import.hx` (I still vote for the leading underscore to avoid case conflicts with `Import.hx` and showing it first in the list of files)

**Comment by Simn** [comment]
I don't have anything technical against the leading underscore, I just find it aesthetically unpleasing.

**Comment by nadako** [comment]
I don't like it aesthetically either, to be honest, but I can't think of an easy alternative to make it easily distinguishable from normal modules.

**Comment by yar3333** [comment]
Don't sure what my opinion is important, but I can't be silent: originally default.hx may be a good idea, but there are many questions come with it. I think intoducing that feature make the code much more hard-to-understand. Right now, we can collect all imports into one file and include it in each source file via "using". I think, one line of code in each file is not a big work to prevent breaking mind during reading the code.

**Comment by nadako** [comment]
I find this feature similar to global namespace static extensions in C# (you put your static extensions into the root ns and you get your extension method available everywhere), it's not ming breaking at all and in our case it's much more explicit.

**Comment by Simn** [comment]
@yar3333: I used to share these exact concerns and in fact I still think they have some merit. I could live with it though because 1. the source of possible input is still finite (one defaults.hx per package involved) and 2. the Haxe compiler has decent display option for resolving types and fields now.

**Comment by ncannasse** [comment]
Regarding the naming I think that `import.hx` is not a bad idea : since it's a keyword in Haxe, it looks indeed like a special file.

**Comment by nadako** [comment]
@ncannasse Still we can't have both `import.hx` and `Import.hx` on Windows. What do you think about  the underscore prefixing idea (see above)?

**Comment by deathbeam** [comment]
Same casing issue happens on Mac OS. And what about naming this file as top-level package? So you will have package for example this structure:

* mylibrary/
  * Main.hx
  * Abc.hx
  * Def.hx

and when you will add `defaults`, it will become

* mylibrary/
  * Mylibrary.hx
  * Main.hx
  * Abc.hx
  * Def.hx

But I am not sure about code with empty package.

**Comment by ncannasse** [comment]
I prefer to "reserve" Import.hx than having an underscore

**Comment by nadako** [comment]
I don't know, this looks like an unfortunate limitation, and I would be very surprised if there's no Import.hx modules in already existing user's code.

**Comment by back2dos** [comment]
I have one. Personally, I prefer renaming it, rather than having to make ASCII art. Not that I don't like ASCII art, but I prefer to do it of my own free will, and not to please the compiler :D

There are other issues to consider, like the precedence of global using and resolve. Currently, `using` takes precedence over `resolve` [as shown here](http://try.haxe.org/#065f7):

```haxe
using Test;

class Test implements Dynamic<String> {
    function new() {}
    function resolve(name:String) return name;
    static function foo(t:Test) {}
    static function main() {
        $type(new Test().bar);//String
        $type(new Test().foo);//Void->Void
    }
}
```

That makes absolute sense. The question is what to do with global using. If it has precedence, then adding it to a whole directory risks clashing with an identifier that was supposed to be handled by `resolve`. Conversely, if you resolve has precedence, then there is no way to apply global using to types that have `resolve`.

I am all for the former option, as the latter kind of mutilates the whole feature. 

Either way, I think we have to accept that this feature will initially blow up in a few people's face. Given just how much convenience it adds (while still remaining explicit), I think it's well worth that.

**Comment by nadako** [comment]
I don't use resolve, but it seems kind of wrong that `using` takes precedence over `resolve`, why is it like that?

**Comment by nadako** [comment]
Okay, after all it's fine with me to reserve lowercase `import.hx` as I don't have a module named `Import.hx`, and it's not a breaking change :)

**Comment by Simn** [comment]
> I don't use resolve, but it seems kind of wrong that using takes precedence over resolve, why is it like that?

If it didn't you effectively couldn't use static extensions at all on a class that has "resolve". I think the order makes sense and the collateral damage is acceptable.

**Comment by fponticelli** [comment]
Why not using an hxml switch?

-imports myspecialmodulemix

For libraries it could be part of extraParams.hxml

**Comment by nadako** [comment]
@fponticelli see discussion about global usings

**Comment by nadako** [comment]
The `defaults_hx` branch is now using `import.hx` file and has enough tests. Anything else to do before merging that?

**Comment by Simn** [comment]
I'm still nervous about the compilation cache, but I'm sure Nicolas will complain soon enough if there are issues.

I'll squash the commits and push it to development.

**Comment by nadako** [comment]
:thumbsup::thumbsup::thumbsup:

**Comment by ncannasse** [comment]
Ok, we need to speed up this a bit: Common.unique_full_path uses a system call which can be expensive when done a lot, and we're sure to do it K times for each module loaded, where K is the package length.

**Comment by ncannasse** [comment]
Also, could it be imports_cache instead of parser_cache?
Thinks seems okay from a quick view wrt compilation server interaction.

**Comment by Simn** [comment]
We do it exactly once per directory at the moment. That's why I added the `parser_cache`.

**Comment by ncannasse** [comment]
once per candidate, so it's K right ?

**Comment by Simn** [comment]
It's not per-module, that's the point.

Regarding the compilation server I can identify three issues:

1. Changing a import.hx: This should invalidate all dependent modules. Can probably be addressed by creating a fake module for the defaults or something.
2. Adding a import.hx: That seems pretty rough, it would have to invalidate all affected modules, but we don't really have that information.
3. Changing a compiler flag: I have no idea if that's an issue if people have stuff like `#if debug using Something;` in their import.hx.

**Comment by ncannasse** [comment]
Compiler cache:
1. right, this can be fixed by adding a dependency, see (2)
2. indeed, but if we create a fake "import" module that is listed as dependency then the compiler cache should handle it as well, we already do this for other modules (see check_module_path)
3. not a problem, different compiler flags will result in different compiler caches anyway

Performances: it still seems to me we're doing several unique_full_path per module loaded even (especially) if there's no imports.hx, which is pretty heavy and should be implemented differently, by doing it only once (they should all have the same prefix since it's file-based and not package-based right ?)

**Comment by nadako** [comment]
We can split off the last part from the module's `unique_file_path` and just add `/import.hx` (or `\import.hx` on windows), to avoid fs query, right?

**Comment by Simn** [comment]
Now I see what you mean, I forgot about that part.

I don't see why we need `unique_file_path` at all: It is already applied to `m_extra.m_file` so we only have to append paths properly.

**Comment by Simn** [comment]
Regarding 2: The problem when adding a new import.hx is that the affected modules don't know that they are affected.

**Comment by ncannasse** [comment]
We need to do the following:
- for each import.hx found, create a module for it and register it as a "real" module. Since it only contains import/using, this should not affect codegen. It will also enable to track file changes
- when a module is found to use one or several import modules, we need to add them to its dependency list

I think this will still not cover the addition of a new import module that would override another one, I just realized that in the dependency list we don't store if module/type was loaded without its full package so we already have the problem when loading types in such way:

```
A.hx
  |- b
      | - B.hx reference type "A"
// cache will not detect that adding b/A.hx requires recompiling B.hx
```

**Comment by Simn** [comment]
Funny, I just wanted to ask about that exact case.

Note that I'm not talking about overriding import.hx files. My point is that when you have a complete application in your cache and add an import.hx somewhere, nothing is gonna happen.

**Comment by ncannasse** [comment]
Right, that would require before each reused module to check if an import.hx file was added.

**Comment by ncannasse** [comment]
Anything you need me for on this one @Simn? There's some lengthy discussion about extra compilation server checks but it seems quite orthogonal to the imports.hx implementation. Should we close and open another one specific to this? Or do you need help to register imports.hx as a proper module (if not yet done)?

**Comment by Simn** [comment]
I'll look into it. I think I still didn't fix the path issue anyway.

**Comment by Simn** [comment]
Alright, I think I've addressed this. There's a bit of a hack because we have to lookup the import.hx module if it exists already. The lookup table indexes using the module path, but for import.hx we have to lookup the file path. Instead of introducing another lookup table I'm using the file path as the module name, so you get for instance `([],"C:\\Github\haxe\std\import.hx")`. I *think* this should be safe?

Other than that I didn't know how to test this properly, but the .dependants.dump looks good:

```
c:\users\simn\documents\testmisc\src\import.hx:
	c:\users\simn\documents\testmisc\src\testext.hx
	c:\users\simn\documents\testmisc\src\main.hx
	c:\users\simn\documents\testmisc\src\b.hx
	c:\users\simn\documents\testmisc\src\a.hx
c:\github\haxe\std\js\_std\import.hx:
	c:\github\haxe\std\js\_std\type.hx
	c:\github\haxe\std\js\_std\string.hx
	c:\github\haxe\std\js\_std\std.hx
	c:\github\haxe\std\js\_std\reflect.hx
	c:\github\haxe\std\js\_std\math.hx
	c:\github\haxe\std\js\_std\hxoverrides.hx
	c:\github\haxe\std\js\_std\date.hx
	c:\github\haxe\std\js\_std\array.hx
c:\github\haxe\std\js\import.hx:
	c:\github\haxe\std\js\lib.hx
	c:\github\haxe\std\js\error.hx
	c:\github\haxe\std\js\boot.hx
c:\github\haxe\std\import.hx:
	c:\github\haxe\std\js\lib.hx
	c:\github\haxe\std\js\error.hx
	c:\github\haxe\std\js\boot.hx
	c:\github\haxe\std\haxe\posinfos.hx
	c:\github\haxe\std\haxe\log.hx
	c:\github\haxe\std\haxe\enumtools.hx
	c:\github\haxe\std\haxe\constraints.hx
	c:\github\haxe\std\stdtypes.hx
	c:\github\haxe\std\intiterator.hx
	c:\github\haxe\std\enumvalue.hx
	c:\github\haxe\std\enum.hx
	c:\github\haxe\std\class.hx
c:\github\haxe\std\haxe\import.hx:
	c:\github\haxe\std\haxe\posinfos.hx
	c:\github\haxe\std\haxe\log.hx
	c:\github\haxe\std\haxe\enumtools.hx
	c:\github\haxe\std\haxe\constraints.hx
```

I'll be optimistic and consider this closed!
